services/text_extractor_service.py

```diff
@@ -14,17 +15,31 @@ class TextExtractorService:
     def __init__(self, db, executor):
         self.executor = executor
         self.stored_recipes_service = StoredRecipesService(db)
+        # You can change maxmemory=100mb and maxmemory-policy=allkeys-lfu from redis-cli.
+        self.redis = redis.Redis(host='redis-10060.c135.eu-central-1-1.ec2.cloud.redislabs.com', port=10060,
+                                 username='default', password='CWqOy4UlreQiOgo8lenpSc6ZMnCvTA8x', decode_responses=True)
 
     def find_recipe_in_url(self, form):
         url = form['url']
         instructions = form['instructions'].lower() == "true"
         ingredients = form['ingredients'].lower() == "true"
         try:
+            if self.redis.hexists(name="frequently_used_recipes", key=url):
+                return jsonify(json.loads(self.redis.hget(name="frequently_used_recipes", key=url)))
             if self.stored_recipes_service.url_in_db(url):
-                return jsonify(self.stored_recipes_service.get_recipe_from_db(url, ingredients, instructions))
+                recipe = self.stored_recipes_service.get_recipe_from_db(url, ingredients, instructions)
+                if self.redis.hset(name="frequently_used_recipes", key=url, value=json.dumps(recipe)):
+                    logging.info(
+                        "cached recipe from url:\n{}".format(url)
+                    )
+                return jsonify(recipe)
```

Comment:
2 things:
1- what happens if we can't cache it for some reason? I think a warning will be appropriate.
2- it's a bit of a duplicated code, can you think of a way to generalize it?

Reply:
I think now it is generalized for most future caching needs